lib/github-pages/configuration.rb

```diff
@@ -0,0 +1,115 @@
+    DEFAULT_PLUGINS = %w(
+      jekyll-coffeescript
+      jekyll-gist
+      jekyll-github-metadata
+      jekyll-paginate
+      jekyll-textile-converter
+    ).freeze
+
+    # Plugins allowed by GitHub Pages
+    PLUGIN_WHITELIST = %w(
+      jekyll-coffeescript
+      jekyll-feed
+      jekyll-gist
+      jekyll-github-metadata
+      jekyll-mentions
+      jekyll-paginate
+      jekyll-redirect-from
+      jekyll-seo-tag
+      jekyll-sitemap
+      jekyll-textile-converter
+      jemoji
+    ).freeze
+
+    # Default, user overwritable options
+    DEFAULTS = {
+      "jailed"   => false,
+      "gems"     => DEFAULT_PLUGINS,
+      "kramdown" => {
+        "input"     => "GFM",
+        "hard_wrap" => false
+      }
+    }.freeze
+
+    # Jekyll defaults merged with Pages defaults.
+    MERGED_DEFAULTS = Jekyll::Utils.deep_merge_hashes(
+      Jekyll::Configuration::DEFAULTS,
+      DEFAULTS
+    ).freeze
+
+    # Options which GitHub Pages sets, regardless of the user-specified value
+    #
+    # The following values are also overridden by GitHub Pages, but are not
+    # overridden locally, for practical purposes:
+    # * source
+    # * destination
+    # * jailed
+    # * verbose
+    # * incremental
+    # * GH_ENV
+    OVERRIDES = {
+      "lsi"         => false,
+      "safe"        => true,
+      "plugins"     => SecureRandom.hex,
+      "whitelist"   => PLUGIN_WHITELIST,
+      "highlighter" => "rouge",
+      "kramdown"    => {
+        "template"          => "",
+        "math_engine"       => "mathjax",
+        "syntax_highligher" => "rouge"
+      },
+      "gist"        => {
+        "noscript"  => false
+      }
+    }.freeze
+
+    # These configuration settings have corresponding instance variables on
+    # Jekyll::Site and need to be set properly when the config is updated.
+    CONFIGS_WITH_METHODS = %w(
+      safe lsi highlighter baseurl exclude include future unpublished
+      show_drafts limit_posts keep_files gems
+    ).freeze
+
+    # Given a user's config, determines the effective configuration by building a user
+    # configuration sandwhich with our overrides overriding the user's specified
+    # values which themselves override our defaults.
+    #
+    # Returns the effective Configuration
+    #
+    # Note: this is a highly modified version of Jekyll#configuration
+    def self.effective_config(user_config)
+      # Merge user config into defaults
+      config = Jekyll::Utils.deep_merge_hashes(MERGED_DEFAULTS, user_config)
+        .fix_common_issues
+        .add_default_collections
+
+      # Merge overwrites into user config
+      Jekyll::Utils.deep_merge_hashes! config, OVERRIDES
+
+      # Ensure we have those gems we want.
+      config["gems"] = Array(config["gems"]) | DEFAULT_PLUGINS
+
+      config
+    end
+
+    # Set the site's configuration  Implemented as an `after_reset` hook.
+    def self.set(site)
+      return if site.config["_github_pages_processed"] == true
```

Comment:
What would happen if a malicious user put `_github_pages_processed: true` in their `_config.yml`? Should we either store it as an instance variable (is that safe?) or override the flag when we override the user's config? If the latter, should we use the existing `github` namespace?